tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/GcovJsonParser.java

```diff
@@ -0,0 +1,120 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.coverageoutputgenerator;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.zip.GZIPInputStream;
+import com.google.gson.Gson;
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * A {@link Parser} for gcov intermediate json format introduced in GCC 9.1. See the flag {@code --intermediate-format} in <a
+ * href="https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Invoking-Gcov.html">gcov documentation</a>.
+ */
+public class GcovJsonParser {
+  private static final Logger logger = Logger.getLogger(GcovJsonParser.class.getName());
+  private final InputStream inputStream;
+
+  private GcovJsonParser(InputStream inputStream) {
+    this.inputStream = inputStream;
+  }
+
+  public static List<SourceFileCoverage> parse(InputStream inputStream) throws IOException {
+    return new GcovJsonParser(inputStream).parse();
+  }
+
+  private List<SourceFileCoverage> parse() throws IOException {
+    ArrayList<SourceFileCoverage> allSourceFiles = new ArrayList<>();
+    try (InputStream gzipStream = new GZIPInputStream(inputStream)) {
+      ByteArrayOutputStream contents = new ByteArrayOutputStream();
+      byte[] buffer = new byte[1024];
+      int length;
+      while ((length = gzipStream.read(buffer)) != -1) {
+        contents.write(buffer, 0, length);
+      }
+      Gson gson = new Gson();
+      GcovJsonFormat document = gson.fromJson(contents.toString(), GcovJsonFormat.class);
+      if (!document.format_version.equals("1")) {
+        logger.log(Level.WARNING, "Expect GCov JSON format version 1, got format version " + document.format_version);
+      }
+      for (GcovJsonFile file : document.files) {
+        SourceFileCoverage currentFileCoverage = new SourceFileCoverage(file.file);
+        for (GcovJsonFunction function : file.functions) {
+          currentFileCoverage.addLineNumber(function.name, function.start_line);
+          currentFileCoverage.addFunctionExecution(function.name, function.execution_count);
+        }
+        for (GcovJsonLine line : file.lines) {
+          currentFileCoverage.addLine(line.line_number, LineCoverage.create(line.line_number, line.count, null));
+          for (GcovJsonBranch branch : line.branches) {
+            currentFileCoverage.addBranch(line.line_number, BranchCoverage.create(line.line_number, branch.count));
+          }
+        }
+        allSourceFiles.add(currentFileCoverage);
+      }
+    }
+
+    return allSourceFiles;
+  }
+}
+
+// Classes for the Gson data mapper representing the structure of the GCov JSON format
+
+class GcovJsonFormat {
```

Comment:
Also, at least provide a *link* to the documentation for the JSON file format.

Reply:
Would it be acceptable to make them inner classes? I think that separating these 5 small classes that do not have associated methods and can't really be reused in other parts of the codebase into their own file would be close to littering.

Furthermore I'm not following the styleguide w.r.t. field naming as well because the elements in the JSON file use this naming, is that an issue as well?

The GCov manpage contains a thorough explanation of the fomat, however I'm having issues finding a versioned permalink other than their git source: https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/doc/gcov.texi;h=dcdd7831ff063483d43e5347af0b67083c85ecc4;hb=4212a6a3e44f870412d9025eeb323fd4f50a61da#l184 which is only in texinfo format

Reply:
Inner classes are fine. I don't know how strict the style check is on the names. Please add a comment saying that these have to match the upstream definitions, so they don't get 'auto-corrected' on import.